libexec/rtld-elf/aarch64/reloc.c

```diff
@@ -828,14 +1092,22 @@ allocate_initial_tls(Obj_Entry *objs)
 	tls_static_space = tls_last_offset + tls_last_size +
 	    RTLD_STATIC_TLS_EXTRA;
 
+#if defined(__CHERI_PURE_CAPABILITY__) && defined(RTLD_SANDBOX)
+	asm ("msr	rctpidr_el0, %0\n" :: "C" (allocate_tls(objs, NULL, TLS_TCB_SIZE, TLS_TCB_ALIGN)));
+#else
 	_tcb_set(allocate_tls(objs, NULL, TLS_TCB_SIZE, TLS_TCB_ALIGN));
+#endif
 }
 
 void *
 __tls_get_addr(tls_index* ti)
 {
 	uintptr_t **dtvp;
 
+#if defined(__CHERI_PURE_CAPABILITY__) && defined(RTLD_SANDBOX)
+	asm ("mrs	%0, rctpidr_el0" : "=C" (dtvp));
```

Comment:
Why not just change _tcb_get/set to do this? You're also a little inconsistent about dtvp vs tcb; above you set it to the tcb, here you get it as dtvp. In practice those are the same because tcb_dtv is the first member, but you shouldn't be assuming that.

Reply:
I think I did this because `__tcb_get/set` would change the Executive mode registers instead of the Restricted mode ones.

Reply:
Good idea to just change `__tcb_get/set`. But these code are in a system header. Which Makefile should I add the `RTLD_SANDBOX` compiler flag to?

Reply:
You don't, just have the header check defined(IN_RTLD) && defined(RTLD_SANDBOX)